### PR TITLE
Experimental support for aliasscopes

### PR DESCRIPTION
**Pull request by vchuravy**
@mbauman the first commit fixes the seqfault you saw. If broadcast is going to use these I figured we should put them into a better place and mark them clearly as experimental.

Test should probably be similar to the ones in https://github.com/JuliaLang/julia/pull/31347

### CODE REVIEW

**Comment by vtjnash** [comment]
Should we put this in `Future`?

**Comment by mbauman** [comment]
`Future` doesn't really make sense for two reasons — one is that it's not certain it'll be in the future, and two is that I want to use it for broadcast to fix #28126.

**Comment by vchuravy** [comment]
Okay I added some simple tests. @mbauman and @YingboMa if you have some more ideas for which cases you care about that would be great.

Jeff/Keno are we okay with adding this implementation as is, I previously didn't do so since all my use-cases were outside of base.

File: src/codegen.cpp [hunk 1]
```diff
@@ -6088,7 +6088,12 @@ static std::unique_ptr<Module> emit_function(
                 } else if (expr->head == popaliasscope_sym) {
                     scope_stack.pop_back();
                     scope_list_stack.pop_back();
-                    current_aliasscope = scope_list_stack.back();
+                    if (scope_list_stack.empty()) {
+                        // XXX: should we error here?
```
Comment (vchuravy) [inline src/codegen.cpp]:
```suggestion
```

**Comment by YingboMa** [comment]
Unfortunately, this PR + https://github.com/JuliaLang/julia/commit/d839bc53432ad9113615f3d18a6209364c0666ad still does quite solve 28126, as
```julia
broadcast_unalias(dest, src) = dest === src ? src : constify(unalias(dest, src))
```
is not type stable.

**Comment by vchuravy** [comment]
Added docstrings that clarify that `Const` and `@aliasscope` are experimental.

**Comment by vchuravy** [comment]
Only 50ec39c should be backported to 1.2

**Comment by vchuravy** [comment]
This looks like a missing root, anyone seen this before?

```
error during bootstrap:
GC error (probable corruption) :
Allocations: 96441027 (Pool: 96436176; Big: 4851); GC: 581
<?#14EF9F80::00000000>
0B37E020: Queued root: 14A62090 :: 059209D0 (bits: 3)
        of type Array{Any, 1}
0B37E02C: Queued root: 14B2F700 :: 05920790 (bits: 3)
        of type Core.TypeMapEntry
0B37E038: Queued root: 13DEB3A0 :: 059209D0 (bits: 3)
        of type Array{Any, 1}
0B37E044: Queued root: 148F9140 :: 059209D0 (bits: 3)
        of type Array{Any, 1}
0B37E050: Queued root: 142F24A0 :: 059209D0 (bits: 3)
        of type Array{Any, 1}
0B37E05C: Queued root: 13DD22C0 :: 059209D0 (bits: 3)
        of type Array{Any, 1}
0B37E068: Queued root: 143EDA00 :: 059209D0 (bits: 3)
        of type Array{Any, 1}
0B37E074: Queued root: 14307BC0 :: 059209D0 (bits: 3)
        of type Array{Any, 1}
0B37E080: Queued root: 136C0880 :: 059209D0 (bits: 3)
        of type Array{Any, 1}
0B37E08C: Queued root: 066E82E0 :: 059209D0 (bits: 3)
        of type Array{Any, 1}
0B37E098:  r-- Stack frame 00C91400 -- 0 of 2 (direct)
Unknown pc 6FC7AF10 --- ABORTING !!!
This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
signal (22): SIGABRT
in expression starting at none:0
raise at C:\Windows\SYSTEM32\msvcrt.dll (unknown line)
abort at C:\Windows\SYSTEM32\msvcrt.dll (unknown line)
gc_assert_datatype_fail at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\gc.c:1522
gc_mark_loop at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\gc.c:2409
_jl_gc_collect at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\gc.c:2715
jl_gc_collect at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\gc.c:2886
jl_gc_pool_alloc at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\gc.c:1102 [inlined]
jl_gc_alloc_ at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\julia_internal.h:274 [inlined]
jl_gc_alloc at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\gc.c:2923
_new_array_ at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\array.c:100
_new_array at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\array.c:160 [inlined]
jl_alloc_array_1d at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\array.c:420
anymap at .\compiler\utilities.jl:39
argtypes_to_type at .\compiler\typeutils.jl:46 [inlined]
abstract_call at .\compiler\abstractinterpretation.jl:800
abstract_call at .\compiler\abstractinterpretation.jl:598
abstract_eval_call at .\compiler\abstractinterpretation.jl:837
abstract_eval at .\compiler\abstractinterpretation.jl:907
typeinf_local at .\compiler\abstractinterpretation.jl:1164
typeinf_nocycle at .\compiler\abstractinterpretation.jl:1220
typeinf at .\compiler\typeinfer.jl:12
abstract_call_method_with_const_args at .\compiler\abstractinterpretation.jl:222
abstract_call_gf_by_type at .\compiler\abstractinterpretation.jl:115
abstract_call at .\compiler\abstractinterpretation.jl:808
abstract_call at .\compiler\abstractinterpretation.jl:598
abstract_eval_call at .\compiler\abstractinterpretation.jl:837
abstract_eval at .\compiler\abstractinterpretation.jl:907
typeinf_local at .\compiler\abstractinterpretation.jl:1164
typeinf_nocycle at .\compiler\abstractinterpretation.jl:1220
typeinf at .\compiler\typeinfer.jl:12
abstract_call_method_with_const_args at .\compiler\abstractinterpretation.jl:222
abstract_call_gf_by_type at .\compiler\abstractinterpretation.jl:115
abstract_call at .\compiler\abstractinterpretation.jl:808
abstract_call at .\compiler\abstractinterpretation.jl:598
abstract_eval_call at .\compiler\abstractinterpretation.jl:837
abstract_eval at .\compiler\abstractinterpretation.jl:907
typeinf_local at .\compiler\abstractinterpretation.jl:1164
typeinf_nocycle at .\compiler\abstractinterpretation.jl:1220
typeinf at .\compiler\typeinfer.jl:12
abstract_call_method_with_const_args at .\compiler\abstractinterpretation.jl:222
abstract_call_gf_by_type at .\compiler\abstractinterpretation.jl:115
abstract_call at .\compiler\abstractinterpretation.jl:808
abstract_call at .\compiler\abstractinterpretation.jl:598
abstract_eval_call at .\compiler\abstractinterpretation.jl:837
abstract_eval at .\compiler\abstractinterpretation.jl:907
typeinf_local at .\compiler\abstractinterpretation.jl:1164
typeinf_nocycle at .\compiler\abstractinterpretation.jl:1220
typeinf at .\compiler\typeinfer.jl:12
abstract_call_method_with_const_args at .\compiler\abstractinterpretation.jl:222
abstract_call_gf_by_type at .\compiler\abstractinterpretation.jl:115
abstract_call at .\compiler\abstractinterpretation.jl:808
abstract_call at .\compiler\abstractinterpretation.jl:598
abstract_eval_call at .\compiler\abstractinterpretation.jl:837
abstract_eval at .\compiler\abstractinterpretation.jl:907
typeinf_local at .\compiler\abstractinterpretation.jl:1150
typeinf_nocycle at .\compiler\abstractinterpretation.jl:1220
typeinf at .\compiler\typeinfer.jl:12
typeinf_edge at .\compiler\typeinfer.jl:482
abstract_call_method at .\compiler\abstractinterpretation.jl:369
abstract_call_gf_by_type at .\compiler\abstractinterpretation.jl:92
abstract_call at .\compiler\abstractinterpretation.jl:808
abstract_call at .\compiler\abstractinterpretation.jl:598
abstract_eval_call at .\compiler\abstractinterpretation.jl:837
abstract_eval at .\compiler\abstractinterpretation.jl:907
typeinf_local at .\compiler\abstractinterpretation.jl:1164
typeinf_nocycle at .\compiler\abstractinterpretation.jl:1220
typeinf at .\compiler\typeinfer.jl:12
typeinf_ext at .\compiler\typeinfer.jl:568
typeinf_ext at .\compiler\typeinfer.jl:599
unknown function (ip: 09F4048D)
jl_apply_generic at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\gf.c:2191
jl_apply at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\julia.h:1604 [inlined]
jl_type_infer at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\gf.c:207
jl_compile_method_internal at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\gf.c:1773
jl_apply_generic at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\gf.c:2196
jl_apply at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\julia.h:1604 [inlined]
jl_call2 at /home/Administrator/buildbot/worker/package_win32/build/src/home/Administrator/buildbot/worker/package_win32/build/src\jlapi.c:227
exec_program at /home/Administrator/buildbot/worker/package_win32/build/ui/home/Administrator/buildbot/worker/package_win32/build/ui\repl.c:45
true_main at /home/Administrator/buildbot/worker/package_win32/build/ui/home/Administrator/buildbot/worker/package_win32/build/ui\repl.c:108
wmain at /home/Administrator/buildbot/worker/package_win32/build/ui/home/Administrator/buildbot/worker/package_win32/build/ui\repl.c:217
__tmainCRTStartup at /usr/src/debug/mingw64-i686-runtime-6.0.0-1/usr/src/debug/mingw64-i686-runtime-6.0.0-1/crt\crtexe.c:334
BaseThreadInitThunk at C:\Windows\SYSTEM32\KERNEL32.DLL (unknown line)
RtlInitializeExceptionChain at C:\Windows\SYSTEM32\ntdll.dll (unknown line)
RtlInitializeExceptionChain at C:\Windows\SYSTEM32\ntdll.dll (unknown line)
Allocations: 96441027 (Pool: 96436176; Big: 4851); GC: 581
*** This error might be fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[1]: Leaving directory '/home/Administrator/buildbot/worker/package_win32/build'
```

**Pull request merged**